pika/channel.py

```diff
@@ -283,7 +352,9 @@ def basic_get(self, callback=None, queue='', no_ack=False):
 
         """
         self._validate_channel_and_callback(callback)
+        # TODO Is basic_get meaningful when callback is None?
         self._on_getok_callback = callback
+        # TODO Strangely, not using _rpc for the synchronous Basic.Get
```

Comment:
@gmr, regarding

> IIRC _rpc doesn't handle the full Basic.GetOk method, header, and body frames.

Good point about how pika's `_rpc` implementation works at present. However, since `Basic.Get` is defined by AMQP as synchronous, it should be treated like all the other synchronous requests and not get ahead of the proverbial queue. I had a long back-and-forth about the semantics of `Basic.Get` with @michaelklishin (pivotal/RabbitMQ), and he beat into me that we should treat `Basic.Get` like all the other synchronous requests, and made a believer out of me.

An additional important benefit of treating `Basic.Get` synchronously is that pika would then be able to allow multiple outstanding `Basic.Get` requests on different or same queue on the same channel and be able to do the right thing. Pika's present non-synchronous implementation of `Basic.Get` breaks when you attempt to make an additional `basic_get` call before the prior one completes (even if the second request is against a different channel). This is because there is only one `Channel._on_getok_callback` instance member, and `Channel` can't properly handle multiple outstanding `Basic.Get` requests this way.